Refactor reward payout code - part 5

HaoranYi opened the pull request:
#### Problem

break up of #31143

(prepare for epoch v2 works)

- remove stake_history param.


#### Summary of Changes

part 5: call member fn to get stake_history

Fixes #
<!-- OPTIONAL: Feature Gate Issue: # -->
<!-- Don't forget to add the "feature-gate" label -->

jeffwashington commented:
I think @behzadnouri needs to review this.

HaoranYi commented on runtime/src/bank.rs:
```diff
@@ -2873,7 +2872,6 @@ impl Bank {
         update_rewards_from_cached_accounts: bool,
     ) {
-        let stake_history = self.stakes_cache.stakes().history().clone();
         let vote_with_stake_delegations_map = self.load_vote_and_stake_accounts(
```
Yes. make sure that stake_history doesn't have invalid items.

behzadnouri commented:
> When stake cache is inconsistent with accounts_db, loading stake_history from the cache, before actual loading vote/stake accounts from accounts db (which will pruning invalid cache keys), can lead to invalid stake/vote accounts to participate into later rewarding code. This is incorrect. With this change, stake_history are loaded correctly after pruning invalid cache keys.

1. If you believe that this commit is actually changing anything, then it is a consensus change and you need to feature-gate it. Please be more careful about changes impacting consensus.

2. I don't understand how pruning invalid cache keys changes `stake_history`. `handle_invalid_keys` does not seem like changing `stake_history`. `stake_history` is only updated in `activate_epoch` which is invoked _before_ rewards calculation code is called.

3. If you see invalid stake/vote accounts in the cache, I think the highest priority thing to do is to debug those inconsistencies and figure out how they creep in. A change like this will only silent and suppress the issue without actually fixing the root cause.

HaoranYi commented:
> I don't know. The initial PR description that is now removed (also in #31143) claims that it does change the computation. So that was a question how does it change.

I updated the description and closed original PR. Also, updated the commit comments.

behzadnouri commented:
> I updated the description and closed original PR. Also, updated the commit comments.

The issue is not the PR description.
The issue is if the change is introducing a change in the resulting computation or not; because then that is a consensus change.

Initially you were saying that:
> This change fixes a potential subtle bug. When stake cache is inconsistent with accounts_db, loading stake_history from the cache, before actual loading vote/stake accounts from accounts db (which will pruning invalid cache keys), can lead to invalid stake/vote accounts to participate into later rewarding code. This is incorrect. With this change, stake_history are loaded correctly after pruning invalid cache keys.

Which was claiming it does introduce a change in the result. Was that wrong? are you sure now this does not change anything?

> Pure refactor. no functional change.

What is the motivation of this change then? i.e. why is this better than before?

HaoranYi commented:
> > I updated the description and closed original PR. Also, updated the commit comments.
> 
> The issue is not the PR description. The issue is if the change is introducing a change in the resulting computation or not; because then that is a consensus change.
> 
> Initially you were saying that:
> 
> > This change fixes a potential subtle bug. When stake cache is inconsistent with accounts_db, loading stake_history from the cache, before actual loading vote/stake accounts from accounts db (which will pruning invalid cache keys), can lead to invalid stake/vote accounts to participate into later rewarding code. This is incorrect. With this change, stake_history are loaded correctly after pruning invalid cache keys.
> 
> Which was claiming it does introduce a change in the result. Was that wrong? are you sure now this does not change anything?

Yes, that remark is incorrect. Therefore, it was removed. 
Yes, I think so. I have reviewed all place that stake_history was updated. Only activate_epoch update stake_history. And that call is outside of reward paying code. Maybe worth another pair of eyes to look at it too?

> > Pure refactor. no functional change.
> 
> What is the motivation of this change then? i.e. why is this better than before?
Better and cleaner code. less fn parameter, less complexity, less dependency. More self-contained. Make it easier to call for future works. Other caller for those fns don't need to worry about passing the history params.

behzadnouri commented:
> > > Pure refactor. no functional change.
> > 
> > What is the motivation of this change then? i.e. why is this better than before? 
>
> Better and cleaner code. less fn parameter, less complexity, less dependency. More self-contained. Make it easier to call for future works. Other caller for those fns don't need to worry about passing the history params.

Frankly, I don't think this is better code.
The old code is explicitly passing the necessary data.
Your changes are obscuring this and hiding it from the caller.
One way this can introduce a bug is if in some future code changes, someone modifies `self.stake_history` between the call to the two functions, unaware that the two computations depend on that value.

HaoranYi commented:
> > > > Pure refactor. no functional change.
> > > 
> > > 
> > > What is the motivation of this change then? i.e. why is this better than before?
> > 
> > 
> > Better and cleaner code. less fn parameter, less complexity, less dependency. More self-contained. Make it easier to call for future works. Other caller for those fns don't need to worry about passing the history params.
> 
> Frankly, I don't think this is better code. The old code is explicitly passing the necessary data. Your changes are obscuring this and hiding it from the caller. One way this can introduce a bug is if in some future code changes, someone modifies `self.stake_history` between the call to the two functions, unaware that the two computations depend on that value.

Fine. We can leave this change out.
